model/riscv_step.sail

```diff
@@ -8,8 +8,113 @@
 
 /* The emulator fetch-execute-interrupt dispatch loop. */
 
-/* returns whether to increment the step count in the trace */
-function step(step_no : int) -> bool = {
+register step_state : StepState
+
+// Checks if the retire returned by the execution of an instruction
+// counts as a step.
+function does_executed_retire_step(r : Retired(Retire_Failure)) -> bool =
+  match r {
+    RETIRE_OK()                        => true,
+    RETIRE_FAIL(Illegal_Instruction()) => true,
+    RETIRE_FAIL(Memory_Exception(_))   => true,
+    RETIRE_FAIL(Trap(_))               => true,
+
+    // do not step if a wait instruction executed.
+    RETIRE_FAIL(Wait_For_Interrupt())  => false,
+
+    // retires from extensions
+    RETIRE_FAIL(Ext_ControlAddr_Check_Failure(_)) => true,
+    RETIRE_FAIL(Ext_DataAddr_Check_Failure(_))    => true,
+    RETIRE_FAIL(Ext_CSR_Check_Failure())          => true,
+    RETIRE_FAIL(Ext_XRET_Priv_Failure(_))         => true,
+}
+
+union Step = {
+  Step_Pending_Interrupt  : (InterruptType, Privilege),
+  Step_Ext_Fetch_Failure  : ext_fetch_addr_error,
+  Step_Execute            : Retired(Retire_Failure),
+  Step_Waiting            : unit
+}
+
+function step_wait_state(step_no : int, exit_wait : bool) -> (Step, bool) = {
+  if shouldWakeForInterrupt() then {
+    if   get_config_print_instr()
+    then print_instr("interrupt exit from WAIT state at PC " ^ BitStr(PC));
+
+    step_state = STEP_ACTIVE;
+    // The waiting instruction retires successfully.  The
+    // pending interrupts will be handled in the next step.
+    (Step_Execute(RETIRE_OK()), true)
+  } else if exit_wait then {
+    // There are no pending interrupts; transition out of the Wait
+    // as instructed.
+    if   get_config_print_instr()
+    then print_instr("forced exit from WAIT state at PC " ^ BitStr(PC));
+
+    step_state = STEP_ACTIVE;
+    // "When TW=1, then if WFI is executed in any
+    // less-privileged mode, and it does not complete within an
+    // implementation-specific, bounded time limit, the WFI
+    // instruction causes an illegal-instruction exception."
```

Comment:
I don't think this is quite right because `exit_wait` is true when it *does* complete. I think when they say "complete" they mean it retires.

Reply:
Maybe we need 3 options for that parameter then:

* Keep waiting
* Retire successfully (spuriously)
* Timeout

Does `wrs` have similar options?